oddt/surface.py

```diff
@@ -0,0 +1,164 @@
+        (usually in protein). Basically reduces the surface to one
+        accesible by other molecules of radius smaller than probe_radius.
+
+    Returns
+    -------
+    verts : numpy array
+        Spatial coordinates for mesh vertices.
+
+    faces : numpy array
+        Faces are defined by referencing vertices from verts.
+    """
+    # Input validation
+    if not isinstance(molecule, oddt.toolkit.Molecule):
+        raise TypeError('molecule needs to be of type oddt.toolkit.Molecule')
+    if not (isinstance(probe_radius, Number) and probe_radius >= 0):
+        raise ValueError('probe_radius needs to be a positive number')
+
+    # Removing waters and hydrogens
+    atom_dict = molecule.atom_dict
+    atom_dict = atom_dict[atom_dict['atomicnum'] != 1]
+    if remove_hoh:
+        if molecule.protein is not True:
+            raise ValueError('Residue names are needed for water removal, '
+                             'molecule.protein property must be set to True')
+        no_hoh = atom_dict['resname'] != 'HOH'
+        atom_dict = atom_dict[no_hoh]
+
+    # Take a molecule's coordinates and atom radii and scale if necessary
+    coords = atom_dict['coords'] * scaling
+    radii = atom_dict['radius'] * scaling
+
+    # More input validation
+    if radii.min() < 1:
+        raise ValueError('Scaling times the radius of the smallest atom must '
+                         'be larger than 1')
+    # Create a ball for each atom in the molecule
+    ball_dict = {radius: ball(radius, dtype=bool) for radius in set(radii)}
+    ball_radii = np.array([ball_dict[radius].shape[0] for radius in radii])
+
+    # Transform the coordinates because the grid starts at (0, 0 ,0)
+    min_coords = np.min(coords, axis=0)
+    max_rad = np.max(ball_radii, axis=0)
+    adjusted = np.round(coords - min_coords + max_rad * 5).astype(np.int64)
+    offset = adjusted[0] - coords[0]
+
+    # Calculate boundries in the grid for each ball.
+    ball_coord_min = (adjusted.T - np.floor(ball_radii / 2).astype(np.int64)).T
+    ball_coord_max = (ball_coord_min.T + ball_radii).T
+
+    # Create the grid
+    grid = np.zeros(shape=ball_coord_max.max(axis=0) + int(8 * scaling), dtype=bool)
+
+    # Place balls in grid
+    for radius, coord_min, coord_max in zip(radii, ball_coord_min, ball_coord_max):
+        grid[coord_min[0]:coord_max[0],
+             coord_min[1]:coord_max[1],
+             coord_min[2]:coord_max[2]] += ball_dict[radius]
+    spacing = (1 / scaling,) * 3
+
+    # Hole-filling with morphological closing
+    grid = binary_closing(grid, ball(probe_radius * 2 * scaling))
+
+    # Marching cubes
+    verts, faces = measure.marching_cubes(grid, level=0, spacing=spacing)[:2]
+
+    # Verts already scaled by the marching cubes function (spacing parameter)
+    # Only need to scale the offset
+    return verts - offset / scaling, faces
+
+
+def find_surface_residues(molecule, max_dist=None, scaling=1.):
```

Comment:
The name of this function is misleading in my opinion. It finds atoms, not residues. Residues are indirectly found this way, as there are `res*` properties for each atom in `atom_dict`.